File: Stack/04.LongestValidParentheses/longestvalidparentheses.js
var longestvalidparentheses = function(s) {
    let longest = 0;
    // 栈用来存储接下来遇到的匹配括号的长度
    let stk = [0];
    for (let i = 0; i < s.length; ++i) {
        if (s[i] === '(') {
            stk.push(0);
        } else {
            if (stk.length > 1) {
                let v = stk.pop();
                // 取出第一个元素
                let lastCount = stk[stk.length - 1];
                stk[stk.length - 1] = lastCount + v + 2;
                longest = Math.max(longest, stk[stk.length - 1]);
            } else {
                stk = [0];
            }
        }
    }
    return longest;
}

File: Stack/04.LongestValidParentheses/longestvalidparentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int longest = 0;
        // 使用 vector 模拟栈，存储接下来遇到的匹配括号的长度
        vector<int> stk;
        stk.push_back(0);
        for (auto ch : s) {
            if (ch == '(') {
                stk.push_back(0);
            }
            else {
                if (stk.size() > 1) {
                    int v = stk.back(); stk.pop_back();
                    int lastCount = stk[stk.size() - 1];
                    stk[stk.size() - 1] = lastCount + v + 2;
                    longest = max(longest, stk[stk.size() - 1]);
                }
                else {
                    stk.clear(); stk.push_back(0);
                }
            }
        }
        return longest;
    }
};